src/ProjectFileIO.cpp

```diff
@@ -2364,219 +2365,80 @@ int64_t ProjectFileIO::GetTotalUsage()
-   int rc;
+   sqlite3_stmt* stmt = nullptr;
 
-   // Get the rootpage for the sampleblocks table.
-   sqlite3_stmt *stmt =
-      conn.Prepare(DBConnection::GetRootPage,
-                    "SELECT rootpage FROM sqlite_master WHERE tbl_name = 'sampleblocks';");
-   if (stmt == nullptr || sqlite3_step(stmt) != SQLITE_ROW)
+   if (blockid == 0)
    {
-      ADD_EXCEPTION_CONTEXT("sqlite3.rc", std::to_string(sqlite3_errcode(conn.DB())));
-      ADD_EXCEPTION_CONTEXT("sqlite3.context", "ProjectGileIO::GetDiskUsage");
+      static const char* statement =
+R"(SELECT 
+	sum(length(blockid) + length(sampleformat) + 
+	length(summin) + length(summax) + length(sumrms) + 
+	length(summary256) + length(summary64k) +
+	length(samples))
+FROM sampleblocks;)";
 
-      return 0;
+      stmt = conn.Prepare(DBConnection::GetAllSampleBlocksSize, statement);
    }
-
-   // And store it in our first stack frame
-   stack.push_back({sqlite3_column_int64(stmt, 0)});
-
-   // All done with the statement
-   sqlite3_clear_bindings(stmt);
-   sqlite3_reset(stmt);
-
-   // Prepare/retrieve statement to read raw database page
-   stmt = conn.Prepare(DBConnection::GetDBPage,
-      "SELECT data FROM sqlite_dbpage WHERE pgno = ?1;");
-   if (stmt == nullptr)
-   {
-      return 0;
-   }
-
-   // Traverse the b-tree until we've visited all of the leaf pages or until
-   // we find the one corresponding to the passed in sample blockid. Because we
-   // use an integer primary key for the sampleblocks table, the traversal will
-   // be in ascending blockid sequence.
-   do
+   else
    {
-      // Acces the top stack frame
-      page &pg = stack.back();
+      static const char* statement =
+R"(SELECT 
+	length(blockid) + length(sampleformat) + 
+	length(summin) + length(summax) + length(sumrms) + 
+	length(summary256) + length(summary64k) +
+	length(samples)
+FROM sampleblocks WHERE blockid = ?1;)";
 
-      // Read the page from the sqlite_dbpage table if it hasn't yet been loaded
-      if (pg.numCells == 0)
-      {
-         // Bind the page number
-         sqlite3_bind_int64(stmt, 1, pg.pgno);
+      stmt = conn.Prepare(DBConnection::GetSampleBlockSize, statement);
+   }
 
-         // And retrieve the page
-         if (sqlite3_step(stmt) != SQLITE_ROW)
+   auto cleanup = finally(
+      [stmt]() {
+         // Clear statement bindings and rewind statement
+         if (stmt != nullptr)
          {
-            // REVIEW: Likely harmless failure - says size is zero on
-            // this error.
-            // LLL: Yea, but not much else we can do.
-            return 0;
+            sqlite3_clear_bindings(stmt);
+            sqlite3_reset(stmt);
          }
+      });
 
-         // Copy the page content to the stack frame
-         memcpy(&pg.data,
-                sqlite3_column_blob(stmt, 0),
-                sqlite3_column_bytes(stmt, 0));
-
-         // And retrieve the total number of cells within it
-         pg.numCells = get2(&pg.data[3]);
```

Comment:
The functions get2, get4, get_varint become unused and might be deleted.

Reply:
Good catch!